File: app/src/main/java/email/schaal/ocreader/ListActivity.java
/*
 * Copyright (C) 2015-2016 Daniel Schaal <daniel@schaal.email>
 *
 * This file is part of OCReader.
 *
 * OCReader is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * OCReader is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with OCReader.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package email.schaal.ocreader;

import android.app.Activity;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;

import androidx.databinding.DataBindingUtil;

import android.os.Bundle;
import android.preference.PreferenceManager;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.google.android.material.snackbar.Snackbar;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.FragmentManager;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import android.view.ActionMode;
import android.view.Menu;
import android.view.MenuItem;

import com.mikepenz.aboutlibraries.Libs;
import com.mikepenz.aboutlibraries.LibsBuilder;

import java.util.Collections;
import java.util.prefs.PreferenceChangeEvent;
import java.util.prefs.PreferenceChangeListener;

import email.schaal.ocreader.database.FeedViewModel;
import email.schaal.ocreader.database.Queries;
import email.schaal.ocreader.database.model.Feed;
import email.schaal.ocreader.database.model.Item;
import email.schaal.ocreader.database.model.TemporaryFeed;
import email.schaal.ocreader.database.model.TreeItem;
import email.schaal.ocreader.databinding.ActivityListBinding;
import email.schaal.ocreader.service.SyncService;
import email.schaal.ocreader.service.SyncType;
import email.schaal.ocreader.view.DividerItemDecoration;
import email.schaal.ocreader.view.FolderBottomSheetDialogFragment;
import email.schaal.ocreader.view.FoldersAdapter;
import email.schaal.ocreader.view.ItemViewHolder;
import email.schaal.ocreader.view.LiveItemsAdapter;

public class ListActivity extends RealmActivity implements ItemViewHolder.OnClickListener, SwipeRefreshLayout.OnRefreshListener, ActionMode.Callback, FoldersAdapter.TreeItemClickListener {
    private static final String TAG = ListActivity.class.getName();

    public static final String LAYOUT_MANAGER_STATE = "LAYOUT_MANAGER_STATE";

    private ActionMode actionMode;
    private ActivityListBinding binding;

    private final BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            final String action = intent.getAction();

            if(action != null && (action.equals(SyncService.SYNC_STARTED) || action.equals(SyncService.SYNC_FINISHED))) {
                final SyncType syncType = SyncType.get(intent.getStringExtra(SyncService.EXTRA_TYPE));
                if(syncType != null) {
                    switch (syncType) {
                        case LOAD_MORE:
                            if (action.equals(SyncService.SYNC_FINISHED)) {
                                //todo: adapter.resetLoadMore();
                            }
                            break;
                        case FULL_SYNC:
                            updateSyncStatus();
                            break;
                    }
                }
            }
        }
    };

    private FeedViewModel feedViewModel;
    private SharedPreferences.OnSharedPreferenceChangeListener preferenceChangeListener;

    private void updateSyncStatus() {
        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this);

        boolean needsUpdate = Preferences.SYS_NEEDS_UPDATE_AFTER_SYNC.getBoolean(sharedPreferences);
        boolean syncRunning = Preferences.SYS_SYNC_RUNNING.getBoolean(sharedPreferences);

        if(needsUpdate) {
            feedViewModel.updateTemporaryFeed(PreferenceManager.getDefaultSharedPreferences(this), true);

            sharedPreferences.edit()
                    .putBoolean(Preferences.SYS_NEEDS_UPDATE_AFTER_SYNC.getKey(), false).apply();
        }

        if (binding.swipeRefreshLayout != null) {
            binding.swipeRefreshLayout.setRefreshing(syncRunning);
        }

        binding.bottomAppbar.getMenu().findItem(R.id.menu_sync).setEnabled(!syncRunning);

        //todo: if(!syncRunning)
        //    adapter.resetLoadMore();
    }

    private LiveItemsAdapter adapter;
    private LinearLayoutManager layoutManager;

    @Override
    protected void onPause() {
        super.onPause();
        LocalBroadcastManager.getInstance(this).unregisterReceiver(broadcastReceiver);
    }

    @Override
    protected void onResume() {
        super.onResume();
        updateSyncStatus();
        LocalBroadcastManager.getInstance(this).registerReceiver(broadcastReceiver, SyncService.syncFilter);
    }

    @Override
    protected void onStart() {
        super.onStart();
        if(!Preferences.hasCredentials(PreferenceManager.getDefaultSharedPreferences(this))) {
            startActivityForResult(new Intent(this, LoginActivity.class), LoginActivity.REQUEST_CODE);
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_list);
        setSupportActionBar(binding.toolbarLayout.toolbar);

        final SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);

        final MenuItem menuItemShowOnlyUnread = binding.bottomAppbar.getMenu().findItem(R.id.menu_show_only_unread);
        menuItemShowOnlyUnread.setChecked(Preferences.SHOW_ONLY_UNREAD.getBoolean(preferences));

        binding.bottomAppbar.setNavigationIcon(R.drawable.ic_folder);
        binding.bottomAppbar.setNavigationOnClickListener(v -> {
            final FragmentManager fm = getSupportFragmentManager();
            final FolderBottomSheetDialogFragment bottomSheetDialogFragment = new FolderBottomSheetDialogFragment();
            bottomSheetDialogFragment.setTreeItemClickListener(this);
            bottomSheetDialogFragment.show(fm, null);
        });
        binding.bottomAppbar.setOnMenuItemClickListener(item -> {
            switch (item.getItemId()) {
                case R.id.menu_settings:
                    final Intent settingsIntent = new Intent(ListActivity.this, SettingsActivity.class);
                    startActivity(settingsIntent);
                    return true;
                case R.id.menu_show_only_unread:
                    final boolean showOnlyUnread = !item.isChecked();
                    preferences.edit().putBoolean(Preferences.SHOW_ONLY_UNREAD.getKey(), showOnlyUnread).apply();
                    item.setChecked(showOnlyUnread);
                    reloadListFragment();

                    return true;
                case R.id.menu_sync:
                    SyncService.startSync(ListActivity.this);
                    return true;
                case R.id.menu_about:
                    showAboutDialog();
                    return true;
                case R.id.menu_mark_all_as_read:
                    Queries.markTemporaryFeedAsRead(getRealm(), null, null);
                    return true;
                case R.id.menu_manage_feeds:
                    startActivityForResult(new Intent(ListActivity.this, ManageFeedsActivity.class), ManageFeedsActivity.REQUEST_CODE);
                    return true;
            }
            return false;
        });

        preferenceChangeListener = (sharedPreferences, key) -> {
            if (Preferences.SHOW_ONLY_UNREAD.getKey().equals(key)) {
                feedViewModel.updateFolders(Preferences.SHOW_ONLY_UNREAD.getBoolean(preferences));
            }
        };

        preferences.registerOnSharedPreferenceChangeListener(preferenceChangeListener);

        binding.swipeRefreshLayout.setColorSchemeResources(R.color.primary);
        binding.swipeRefreshLayout.setOnRefreshListener(this);

        layoutManager = new LinearLayoutManager(this);

        feedViewModel = ViewModelProviders.of(this, new FeedViewModel.FeedViewModelFactory(this)).get(FeedViewModel.class);

        adapter = new LiveItemsAdapter(Collections.emptyList(), this);

        feedViewModel.getItems().observe(this, items -> {
            if(adapter != null) {
                adapter.updateItems(items);
                if(adapter.getSelectedItemsCount() > 0 && actionMode == null) {
                    actionMode = startActionMode(this);
                }
            }
            binding.listviewSwitcher.setDisplayedChild(items.isEmpty() ? 0 : 1);
        });

        feedViewModel.getTemporaryFeed().observe(this, temporaryFeed -> {
            //noinspection ConstantConditions
            getSupportActionBar().setTitle(temporaryFeed.getName());
        });

        binding.itemsRecyclerview.setAdapter(adapter);
        binding.itemsRecyclerview.setLayoutManager(layoutManager);

        binding.itemsRecyclerview.addItemDecoration(new DividerItemDecoration(this, R.dimen.divider_inset));

        if(savedInstanceState != null) {
            layoutManager.onRestoreInstanceState(savedInstanceState.getParcelable(LAYOUT_MANAGER_STATE));
            adapter.onRestoreInstanceState(savedInstanceState);
        }
    }

    private boolean isShowOnlyUnread() {
        return Preferences.SHOW_ONLY_UNREAD.getBoolean(PreferenceManager.getDefaultSharedPreferences(this));
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable(LAYOUT_MANAGER_STATE, layoutManager.onSaveInstanceState());
        adapter.onSaveInstanceState(outState);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }

    private void reloadListFragment() {
        feedViewModel.updateTemporaryFeed(PreferenceManager.getDefaultSharedPreferences(this), true);
        binding.itemsRecyclerview.scrollToPosition(0);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable final Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if(resultCode == Activity.RESULT_OK) {
            switch (requestCode) {
                case LoginActivity.REQUEST_CODE:
                    if (data != null && data.getBooleanExtra(LoginActivity.EXTRA_IMPROPERLY_CONFIGURED_CRON, false)) {
                        Snackbar.make(binding.coordinatorLayout, R.string.updater_improperly_configured, Snackbar.LENGTH_INDEFINITE)
                                .setAction(R.string.more_info, v -> startActivity(data))
                                .setActionTextColor(ContextCompat.getColor(this, R.color.warning))
                                .show();
                    }
                    reloadListFragment();
                    Queries.resetDatabase();
                    SyncService.startSync(this, true);
                    break;
                case ItemPagerActivity.REQUEST_CODE:
                    if(data != null)
                        binding.itemsRecyclerview.smoothScrollToPosition(data.getIntExtra(ItemPagerActivity.EXTRA_CURRENT_POSITION, -1));
                    break;
                case ManageFeedsActivity.REQUEST_CODE:
                    reloadListFragment();
                    break;
            }
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_item_list_top, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch(item.getItemId()) {
            case R.id.menu_account:
                Intent loginIntent = new Intent(ListActivity.this, LoginActivity.class);
                startActivityForResult(loginIntent, LoginActivity.REQUEST_CODE);
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void showAboutDialog() {
        new LibsBuilder()
                .withAboutIconShown(true)
                .withAboutVersionShown(true)
                .withAboutDescription(getString(R.string.about_app, getString(R.string.app_year_author), getString(R.string.app_url)))
                .withAboutAppName(getString(R.string.app_name))
                .withLicenseShown(true)
                .withActivityStyle(Preferences.DARK_THEME.getBoolean(PreferenceManager.getDefaultSharedPreferences(this)) ? Libs.ActivityStyle.DARK : Libs.ActivityStyle.LIGHT_DARK_TOOLBAR)
                .withActivityTitle(getString(R.string.about))
                .withFields(R.string.class.getFields())
                .start(this);
    }

    @Override
    public void onItemClick(Item item, int position) {
        if(actionMode == null) {
            Intent itemActivityIntent = new Intent(this, ItemPagerActivity.class);
            itemActivityIntent.putExtra(ItemPagerActivity.EXTRA_CURRENT_POSITION, position);
            startActivityForResult(itemActivityIntent, ItemPagerActivity.REQUEST_CODE);
        } else {
            adapter.toggleSelection(position);
            if(adapter.getSelectedItemsCount() == 0)
                actionMode.finish();
            else {
                actionMode.setTitle(String.valueOf(adapter.getSelectedItemsCount()));
                actionMode.invalidate();
            }
        }
    }

    @Override
    public void onItemLongClick(Item item, int position) {
        if(actionMode != null || Preferences.SYS_SYNC_RUNNING.getBoolean(PreferenceManager.getDefaultSharedPreferences(this)))
            return;

        adapter.toggleSelection(position);
        actionMode = startActionMode(this);
    }

    @Override
    public void onRefresh() {
        SyncService.startSync(this);
    }

    // TODO: 12/21/19 implement loadmore
    public void onLoadMore(@NonNull TreeItem treeItem) {
        final Number minId = TemporaryFeed.getListTemporaryFeed(getRealm())
                .getItems()
                .where()
                .min(Item.ID);

        // minId is null if there are no feed items in treeItem
        SyncService.startLoadMore(this, treeItem.getId(), minId != null ? minId.longValue() : 0, treeItem instanceof Feed);
    }

    @Override
    public boolean onCreateActionMode(ActionMode mode, Menu menu) {
        mode.getMenuInflater().inflate(R.menu.menu_item_list_action, menu);
        mode.setTitle(String.valueOf(adapter.getSelectedItemsCount()));
        binding.swipeRefreshLayout.setEnabled(false);
        return true;
    }

    @Override
    public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
        int selectedItemsCount = adapter.getSelectedItemsCount();

        // the menu only changes on the first and second selection
        if(selectedItemsCount > 2)
            return false;

        Item firstSelectedItem = adapter.getFirstSelectedItem();

        boolean firstSelectedUnread = firstSelectedItem != null && firstSelectedItem.isUnread();
        menu.findItem(R.id.action_mark_read).setVisible(firstSelectedUnread);
        menu.findItem(R.id.action_mark_unread).setVisible(!firstSelectedUnread);

        boolean firstSelectedStarred = firstSelectedItem != null && firstSelectedItem.isStarred();
        menu.findItem(R.id.action_mark_starred).setVisible(!firstSelectedStarred);
        menu.findItem(R.id.action_mark_unstarred).setVisible(firstSelectedStarred);

        menu.findItem(R.id.action_mark_above_read).setVisible(selectedItemsCount == 1);

        return true;
    }

    @Override
    public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_mark_read:
                Queries.setItemsUnread(getRealm(), false, adapter.getSelectedItems());
                mode.finish();
                return true;
            case R.id.action_mark_unread:
                Queries.setItemsUnread(getRealm(), true, adapter.getSelectedItems());
                mode.finish();
                return true;
            case R.id.action_mark_starred:
                Queries.setItemsStarred(getRealm(), true, adapter.getSelectedItems());
                mode.finish();
                return true;
            case R.id.action_mark_unstarred:
                Queries.setItemsStarred(getRealm(), false, adapter.getSelectedItems());
                mode.finish();
                return true;
            case R.id.action_mark_above_read:
                Queries.markAboveAsRead(getRealm(), feedViewModel.getItems().getValue(), adapter.getSelectedItems()[0].getId());
                mode.finish();
                return true;
        }
        return false;
    }

    @Override
    public void onDestroyActionMode(ActionMode mode) {
        actionMode = null;
        binding.swipeRefreshLayout.setEnabled(true);
        adapter.clearSelection();
    }

    @Override
    public void onTreeItemClick(final TreeItem treeItem) {
        feedViewModel.updateSelectedTreeItem(treeItem);
        reloadListFragment();
    }
}

File: app/src/main/java/email/schaal/ocreader/ListActivity.kt
/*
 * Copyright (C) 2015-2016 Daniel Schaal <daniel@schaal.email>
 *
 * This file is part of OCReader.
 *
 * OCReader is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * OCReader is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with OCReader.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package email.schaal.ocreader

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.os.Bundle
import android.view.ActionMode
import android.view.Menu
import android.view.MenuItem
import androidx.activity.result.contract.ActivityResultContract
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.result.launch
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar.OnMenuItemClickListener
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener
import com.google.android.material.snackbar.Snackbar
import com.mikepenz.aboutlibraries.LibsBuilder
import email.schaal.ocreader.R.string
import email.schaal.ocreader.database.FeedViewModel
import email.schaal.ocreader.database.FeedViewModel.FeedViewModelFactory
import email.schaal.ocreader.database.Queries
import email.schaal.ocreader.database.model.Item
import email.schaal.ocreader.database.model.TemporaryFeed
import email.schaal.ocreader.database.model.TreeItem
import email.schaal.ocreader.databinding.ActivityListBinding
import email.schaal.ocreader.service.SyncResultReceiver
import email.schaal.ocreader.service.SyncType
import email.schaal.ocreader.view.*
import email.schaal.ocreader.view.TreeItemsAdapter.TreeItemClickListener

class ListActivity : AppCompatActivity(), ItemViewHolder.OnClickListener, OnRefreshListener, ActionMode.Callback, TreeItemClickListener {
    private lateinit var bottomMenuClickListener: OnMenuItemClickListener
    private var actionMode: ActionMode? = null
    private lateinit var binding: ActivityListBinding
    private lateinit var adapter: LiveItemsAdapter
    private lateinit var layoutManager: LinearLayoutManager
    private lateinit var syncResultReceiver: SyncResultReceiver
    private val feedViewModel: FeedViewModel by viewModels { FeedViewModelFactory(this) }
    private lateinit var preferenceChangeListener: OnSharedPreferenceChangeListener

    private fun updateSyncStatus(syncRunning: Boolean) {
        if(!syncRunning)
            feedViewModel.updateTemporaryFeed(PreferenceManager.getDefaultSharedPreferences(this), true)

        binding.swipeRefreshLayout.isRefreshing = syncRunning
        binding.bottomAppbar.menu.findItem(R.id.menu_sync).isEnabled = !syncRunning
    }

    override fun onStart() {
        super.onStart()
        if (!Preferences.hasCredentials(PreferenceManager.getDefaultSharedPreferences(this))) {
            getLoginResult.launch()
        }
    }

    private val getSettingsResult = registerForActivityResult(object : ActivityResultContract<Unit, Boolean>() {
        override fun createIntent(context: Context, input: Unit?): Intent {
            return Intent(this@ListActivity, SettingsActivity::class.java)
        }

        override fun parseResult(resultCode: Int, intent: Intent?): Boolean {
            return if (resultCode == Activity.RESULT_OK && intent != null)
                intent.getBooleanExtra(SettingsActivity.EXTRA_RECREATE_ACTIVITY, false)
            else false
        }

    }) { if (it) recreate() }

    private val getManageFeedsResult = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
        if (it.resultCode == Activity.RESULT_OK)
            reloadListFragment()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_list)
        syncResultReceiver = SyncResultReceiver(binding.listviewSwitcher)
        setSupportActionBar(binding.toolbarLayout.toolbar)
        val preferences = PreferenceManager.getDefaultSharedPreferences(this)
        val menuItemShowOnlyUnread = binding.bottomAppbar.menu.findItem(R.id.menu_show_only_unread)
        menuItemShowOnlyUnread.isChecked = Preferences.SHOW_ONLY_UNREAD.getBoolean(preferences)

        bottomMenuClickListener = OnMenuItemClickListener { item: MenuItem ->
            when (item.itemId) {
                R.id.menu_settings -> {
                    getSettingsResult.launch()
                    true
                }
                R.id.menu_show_only_unread -> {
                    val showOnlyUnread = !item.isChecked
                    preferences.edit().putBoolean(Preferences.SHOW_ONLY_UNREAD.key, showOnlyUnread).apply()
                    item.isChecked = showOnlyUnread
                    reloadListFragment()
                    true
                }
                R.id.menu_sync -> {
                    feedViewModel.sync(this, SyncType.FULL_SYNC, syncResultReceiver)
                    true
                }
                R.id.menu_about -> {
                    showAboutDialog()
                    true
                }
                R.id.menu_mark_all_as_read -> {
                    feedViewModel.markTemporaryFeedAsRead()
                    true
                }
                R.id.menu_manage_feeds -> {
                    getManageFeedsResult.launch(Intent(this, ManageFeedsActivity::class.java))
                    true
                } else -> false
            }
        }

        binding.bottomAppbar.apply {
            setOnMenuItemClickListener(bottomMenuClickListener)
            setNavigationIcon(R.drawable.ic_folder)
            setNavigationOnClickListener {
                FolderBottomSheetDialogFragment().apply {
                    setTreeItemClickListener(this@ListActivity)
                }.show(supportFragmentManager, null)
            }
        }

        preferenceChangeListener = OnSharedPreferenceChangeListener { _: SharedPreferences?, key: String ->
            when (key) {
                Preferences.SHOW_ONLY_UNREAD.key -> {
                    feedViewModel.updateFolders(Preferences.SHOW_ONLY_UNREAD.getBoolean(preferences))
                }
                Preferences.SYS_SYNC_RUNNING.key -> {
                    updateSyncStatus(Preferences.SYS_SYNC_RUNNING.getBoolean(preferences))
                }
            }
        }

        preferences.registerOnSharedPreferenceChangeListener(preferenceChangeListener)
        binding.swipeRefreshLayout.setColorSchemeResources(R.color.primary)
        binding.swipeRefreshLayout.setOnRefreshListener(this)
        layoutManager = LinearLayoutManager(this)
        adapter = LiveItemsAdapter(emptyList(), this)
        feedViewModel.items.observe(this, Observer { items: List<Item> ->
            adapter.updateItems(items)
            if (adapter.selectedItemsCount > 0 && actionMode == null) {
                actionMode = startActionMode(this)
            }
            binding.listviewSwitcher.displayedChild = if (items.isEmpty()) 0 else 1
        })
        feedViewModel.temporaryFeed.observe(this, Observer { temporaryFeed: TemporaryFeed -> supportActionBar?.title = temporaryFeed.name })

        feedViewModel.syncStatus.observe(this, Observer { updateSyncStatus(it) })

        binding.itemsRecyclerview.adapter = adapter
        binding.itemsRecyclerview.layoutManager = layoutManager
        binding.itemsRecyclerview.addItemDecoration(DividerItemDecoration(this, R.dimen.divider_inset))
        if (savedInstanceState != null) {
            layoutManager.onRestoreInstanceState(savedInstanceState.getParcelable(LAYOUT_MANAGER_STATE))
            adapter.onRestoreInstanceState(savedInstanceState)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(LAYOUT_MANAGER_STATE, layoutManager.onSaveInstanceState())
        adapter.onSaveInstanceState(outState)
    }

    private fun reloadListFragment() {
        feedViewModel.updateTemporaryFeed(PreferenceManager.getDefaultSharedPreferences(this), true)
        binding.itemsRecyclerview.scrollToPosition(0)
    }

    val getLoginResult = registerForActivityResult(object: ActivityResultContract<Void?, String?>() {
        override fun createIntent(context: Context, input: Void?): Intent {
            return Intent(this@ListActivity, LoginFlowActivity::class.java).apply {
                Preferences.URL.getString(PreferenceManager.getDefaultSharedPreferences(this@ListActivity))?.let {
                    putExtra(LoginFlowActivity.EXTRA_URL, it)
                }
            }
        }

        override fun parseResult(resultCode: Int, intent: Intent?): String? {
            return when(resultCode) {
                Activity.RESULT_OK -> {
                    if(intent?.getBooleanExtra(LoginFlowActivity.EXTRA_IMPROPERLY_CONFIGURED_CRON, false) == true)
                        getString(string.updater_improperly_configured)
                    else
                        null
                }
                else -> {
                    intent?.getStringExtra(LoginFlowActivity.EXTRA_MESSAGE)
                }
            }
        }
    }) { message ->
        message?.let {
            Snackbar.make(binding.coordinatorLayout, it, Snackbar.LENGTH_LONG)
                    .show()
        }
        Queries.resetDatabase()
        feedViewModel.sync(this, SyncType.FULL_SYNC, syncResultReceiver)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_item_list_top, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_account -> {
                val bottomSheetDialogFragment = UserBottomSheetDialogFragment()
                bottomSheetDialogFragment.show(supportFragmentManager, null)

                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    fun onNavigationItemClick(item: MenuItem): Boolean {
        return bottomMenuClickListener.onMenuItemClick(item)
    }

    private fun showAboutDialog() {
        LibsBuilder()
                .withAboutIconShown(true)
                .withAboutVersionShown(true)
                .withAboutDescription(getString(string.about_app, getString(string.app_year_author), getString(string.app_url)))
                .withAboutAppName(getString(string.app_name))
                .withLicenseShown(true)
                .withActivityTitle(getString(string.about))
                .withFields(string::class.java.fields)
                .start(this)
    }

    private val getItemPagerResult = registerForActivityResult(object: ActivityResultContract<Int, Int>() {
        override fun createIntent(context: Context, position: Int): Intent {
            return Intent(this@ListActivity, ItemPagerActivity::class.java).apply {
                putExtra(ItemPagerActivity.EXTRA_CURRENT_POSITION, position)
            }
        }

        override fun parseResult(resultCode: Int, intent: Intent?): Int {
            return if(resultCode == Activity.RESULT_OK && intent != null)
                intent.getIntExtra(ItemPagerActivity.EXTRA_CURRENT_POSITION, -1) else -1
        }

    }) {
        if(it >= 0) binding.itemsRecyclerview.smoothScrollToPosition(it)
    }

    override fun onItemClick(item: Item, position: Int) {
        if (actionMode == null) {
            getItemPagerResult.launch(position)
        } else {
            adapter.toggleSelection(position)
            if (adapter.selectedItemsCount == 0) actionMode?.finish() else {
                actionMode?.title = adapter.selectedItemsCount.toString()
                actionMode?.invalidate()
            }
        }
    }

    override fun onItemLongClick(item: Item, position: Int) {
        if (actionMode != null || feedViewModel.syncStatus.value == true) return
        adapter.toggleSelection(position)
        actionMode = startActionMode(this)
    }

    override fun onRefresh() {
        feedViewModel.sync(this, SyncType.FULL_SYNC, syncResultReceiver)
    }

    override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
        mode.menuInflater.inflate(R.menu.menu_item_list_action, menu)
        mode.title = adapter.selectedItemsCount.toString()
        binding.swipeRefreshLayout.isEnabled = false
        return true
    }

    override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
        val selectedItemsCount = adapter.selectedItemsCount
        // the menu only changes on the first and second selection
        if (selectedItemsCount > 2) return false
        val firstSelectedItem = adapter.firstSelectedItem
        val firstSelectedUnread = firstSelectedItem != null && firstSelectedItem.unread
        menu.findItem(R.id.action_mark_read).isVisible = firstSelectedUnread
        menu.findItem(R.id.action_mark_unread).isVisible = !firstSelectedUnread
        val firstSelectedStarred = firstSelectedItem != null && firstSelectedItem.starred
        menu.findItem(R.id.action_mark_starred).isVisible = !firstSelectedStarred
        menu.findItem(R.id.action_mark_unstarred).isVisible = firstSelectedStarred
        menu.findItem(R.id.action_mark_above_read).isVisible = selectedItemsCount == 1
        return true
    }

    override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_mark_read -> {
                feedViewModel.setItemUnread(false, *adapter.selectedItems)
                mode.finish()
                return true
            }
            R.id.action_mark_unread -> {
                feedViewModel.setItemUnread(true, *adapter.selectedItems)
                mode.finish()
                return true
            }
            R.id.action_mark_starred -> {
                feedViewModel.setItemStarred(true, *adapter.selectedItems)
                mode.finish()
                return true
            }
            R.id.action_mark_unstarred -> {
                feedViewModel.setItemStarred(false, *adapter.selectedItems)
                mode.finish()
                return true
            }
            R.id.action_mark_above_read -> {
                val lastItemId = adapter.selectedItems[0]?.id
                if(lastItemId != null)
                    feedViewModel.markAboveAsRead(feedViewModel.items.value, lastItemId)
                mode.finish()
                return true
            }
        }
        return false
    }

    override fun onDestroyActionMode(mode: ActionMode) {
        actionMode = null
        binding.swipeRefreshLayout.isEnabled = true
        adapter.clearSelection()
    }

    override fun onTreeItemClick(treeItem: TreeItem) {
        feedViewModel.updateSelectedTreeItem(this, treeItem)
        reloadListFragment()
    }

    companion object {
        private val TAG = ListActivity::class.java.name
        const val LAYOUT_MANAGER_STATE = "LAYOUT_MANAGER_STATE"
    }
}